Add Support for Legacy and New Scopes

dshanske opened the pull request:
Right now the plugin completely ignores scopes. This gives it support for current and future scopes.

dshanske committed:
Add Support for Legacy and New Scopes

snarfed commented:
this is really great, i'm glad we're adding scope support. out of curiosity though, why add a filter? what's the use case for other plugins or themes overriding scope auth checks?

dshanske commented:
New scopes.

dshanske commented:
@snarfed Rebased this based on the other merges. I'm willing to remove the filter, but I think it would help optional expansion. Will defer.

snarfed commented:
yeah, let's drop the filter for now. i'd need to hear at least a couple different concrete use cases for extensible scope authorization before it would make sense to me personally, and maybe not even then. auth is important and delicate, and not generally something you want to let any other code, written by anyone, arbitrarily override.

aaronpk commented:
Agreed on removing the filter. Until there is a good use case for a plugin modifying the behavior of the plugin that also wants to provide finer-grained access, there's no need to make that part extensible. Supporting the primary list of scopes create/update/delete/undelete is good for now, and you could also accept "post" if it's requested for legacy clients.

dshanske commented:
@snarfed You suggested that I removed post scope. I didn't, I just moved the check further down into the capability function. It treats post as full permission, and create, update, delete, or undelete as the scopes for the individual functions... although with only a few clients supporting undelete, maybe delete should be the scope for either.

snarfed commented:
true! you're absolutely right, sorry about that. thanks!

snarfed reviewed (approved):
lgtm! just one refactoring suggestion. feel free to merge after that!

dshanske commented:
I am thinking that I might make some other changes now that I am not using a filter. Move the WordPress permission check into the check scope function. May ask for your thoughts once done.

snarfed commented on micropub.php:
```diff
@@ -246,6 +250,19 @@ private static function authorize() {
+	 * @return boolean
+	**/
+	protected static function check_scope( $scope ) {
+		if ( in_array( $scope, static::$scope, true ) || in_array( 'post', static::$scope, true ) ) {
```
very minor nit: this can now be just `return in_array(...) || in_array(...);`

snarfed commented on readme.md:
```diff
@@ -17,6 +17,18 @@ Supports the [full W3C Micropub CR spec](https://www.w3.org/TR/micropub/) as of
+## Scope 
+
+Supports the following [scope](https://indieweb.org/scope) parameters requested by Micropub clients.
+* Post (legacy) - Grants all user delegated access
```
these should probably be lower case, eg `post`, since they're case sensitive.

snarfed commented on readme.txt:
```diff
@@ -25,6 +25,17 @@ Supports the [full W3C Micropub CR spec](https://www.w3.org/TR/micropub/) as of
 
+== Scope ==
+
+Supports the following [scope](https://indieweb.org/scope) parameters requested by Micropub clients.
```
ditto

snarfed reviewed (commented):
looks good! only a couple minor nits.

btw, github evidently doesn't notify a PR's reviewers etc when you just amend an existing commit like this. it also makes reviewing a bit harder, since it's hard to tell just what's changed since you last reviewed. might be one reason to wait on squashing until after reviewing. just fyi.
